File: netlify/functions/fetch-video.js
// We use 'node-fetch' for making server-to-server HTTP requests.
const fetch = require('node-fetch');

/**
 * This is the final, robust serverless function.
 * It acts as a secure backend proxy to the single best public API available.
 * It includes detailed error handling for all known scenarios.
 */
exports.handler = async function(event) {
    // 1. Get the YouTube URL from the request sent by our frontend.
    const { url } = event.queryStringParameters;

    if (!url) {
        return {
            statusCode: 400, // Bad Request
            body: JSON.stringify({ message: 'ไม่พบ URL ของวิดีโอในคำขอ' }),
        };
    }

    // 2. Define the single, industry-standard API endpoint. We commit to this one.
    const API_ENDPOINT = "https://api.cobalt.tools/api/json";

    try {
        // 3. The function makes the request from Netlify's servers.
        const response = await fetch(API_ENDPOINT, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json',
            },
            body: JSON.stringify({ url: url }),
        });

        const data = await response.json();

        // 4. IMPORTANT: We handle all possible statuses from the API.
        // This provides clear feedback to the user instead of a generic error.
        switch (data.status) {
            case 'stream':
            case 'picker':
            case 'success':
                // If successful, send the clean data back to our frontend.
                return {
                    statusCode: 200,
                    headers: { "Content-Type": "application/json" },
                    body: JSON.stringify(data),
                };
            case 'error':
                // If the API itself reports an error (e.g., video is private).
                throw new Error(data.text || 'API ไม่สามารถประมวลผลวิดีโอนี้ได้ (อาจเป็นวิดีโอส่วนตัวหรือมีข้อจำกัด)');
            case 'redirect':
                throw new Error('ลิงก์นี้เป็นลิงก์สำหรับ перенаправление (redirect) และไม่สามารถใช้งานได้');
            default:
                 // For any other unexpected status.
                throw new Error(`พบสถานะที่ไม่รู้จักจากเซิร์ฟเวอร์: ${data.status}`);
        }

    } catch (error) {
        // 5. If any step in the `try` block fails, this block will run.
        console.error("Backend Function Error:", error);
        return {
            statusCode: 502, // Bad Gateway (indicates an issue with the upstream API)
            body: JSON.stringify({ message: error.message || 'เกิดข้อผิดพลาดฝั่งเซิร์ฟเวอร์ในการดึงข้อมูล' }),
        };
    }
};

File: netlify/functions/fetch_video.py
import json
import yt_dlp

def handler(event, context):
    """
    This is the main handler for the Netlify Function, written in Python.
    It directly uses the yt_dlp library to fetch video info.
    """
    try:
        # 1. Get the YouTube URL from the query parameters.
        url = event.get('queryStringParameters', {}).get('url')
        if not url:
            return {
                'statusCode': 400,
                'body': json.dumps({'message': 'ไม่พบ URL ของวิดีโอ'})
            }

        # 2. Set options for yt_dlp to get metadata without downloading.
        ydl_opts = {
            'quiet': True,
            'no_warnings': True,
            'dump_single_json': True,
            'skip_download': True,
        }

        # 3. Run yt_dlp to extract video information.
        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            metadata = ydl.extract_info(url, download=False)

        # 4. Transform the metadata into the format our frontend expects.
        picker_items = []
        for f in metadata.get('formats', []):
            if f.get('vcodec') != 'none' and f.get('acodec') != 'none' and f.get('ext') == 'mp4':
                picker_items.append({
                    'url': f.get('url'),
                    'quality': f.get('format_note') or f'{f.get("height")}p',
                    'type': 'video',
                    'audio': True
                })
        
        # Sort by quality and remove duplicates
        unique_qualities = {}
        for item in sorted(picker_items, key=lambda x: int(x['quality'][:-1]), reverse=True):
            if item['quality'] not in unique_qualities:
                unique_qualities[item['quality']] = item
        
        response_payload = {
            'status': 'picker',
            'title': metadata.get('title'),
            'thumbnail': metadata.get('thumbnail'),
            'picker': list(unique_qualities.values())
        }

        # 5. Return a successful response.
        return {
            'statusCode': 200,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps(response_payload)
        }

    except Exception as e:
        # 6. If any step fails, return a structured error message.
        print(f"Error processing video: {e}")
        return {
            'statusCode': 500,
            'body': json.dumps({'message': f'เกิดข้อผิดพลาดในการประมวลผลวิดีโอ: {str(e)}'})
        }


